staticd/static_bfd.c

```diff
@@ -0,0 +1,789 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA.
+ */
+
+#include <zebra.h>
+
+#include "lib/printfrr.h"
+#include "lib/srcdest_table.h"
+
+#include "staticd/static_routes.h"
+#include "staticd/static_zebra.h"
+
+#include "lib/openbsd-queue.h"
+
+DEFINE_MTYPE_STATIC(STATIC, ROUTE_GROUP, "route group memory usage");
+DEFINE_MTYPE_STATIC(STATIC, ROUTE_GROUP_MEMBER,
+		    "route group member memory usage");
+
+/** Global static BFD integration variables. */
+struct static_bfd_global {
+	/** Static route group list. */
+	struct srglist sbg_srglist;
+};
+
+static struct static_bfd_global sbglobal;
+
+/*
+ * Next hop BFD monitoring settings.
+ */
+static void static_next_hop_bfd_change(struct static_nexthop *sn,
+				       const struct bfd_session_status *bss)
+{
+	switch (bss->state) {
+	case BSS_UNKNOWN:
+		/* FALLTHROUGH: no known state yet. */
+	case BSS_ADMIN_DOWN:
+		/* NOTHING: we or the remote end administratively shutdown. */
+		break;
+	case BSS_DOWN:
+		/* Peer went down, remove this next hop. */
+		zlog_info("%s: next hop is down, remove it from RIB", __func__);
+		sn->path_down = true;
+		static_zebra_route_add(sn->rn, sn->sp, sn->safi, true);
+		break;
+	case BSS_UP:
+		/* Peer is back up, add this next hop. */
+		zlog_info("%s: next hop is up, add it to RIB", __func__);
+		sn->path_down = false;
+		static_zebra_route_add(sn->rn, sn->sp, sn->safi, true);
+		break;
+	}
+}
+
+static void static_next_hop_bfd_updatecb(
+	__attribute__((unused)) struct bfd_session_params *bsp,
+	const struct bfd_session_status *bss, void *arg)
+{
+	static_next_hop_bfd_change(arg, bss);
+}
+
+void static_next_hop_bfd_monitor_enable(struct static_nexthop *sn,
+					const struct lyd_node *dnode)
+{
+	bool use_interface;
+	bool use_profile;
+	bool onlink;
+	bool mhop;
+	int family;
+
+	use_interface = false;
+	use_profile = yang_dnode_exists(dnode, "../profile");
+	onlink = yang_dnode_exists(dnode, "../../onlink")
+		 && yang_dnode_get_bool(dnode, "../../onlink");
+	mhop = yang_dnode_get_bool(dnode, "../multi-hop");
+
+	switch (sn->type) {
+	case STATIC_IPV4_GATEWAY_IFNAME:
+	case STATIC_IPV6_GATEWAY_IFNAME:
+		use_interface = true;
+		/* FALLTHROUGH */
+	case STATIC_IPV4_GATEWAY:
+	case STATIC_IPV6_GATEWAY:
+		if (sn->type == STATIC_IPV4_GATEWAY
+		    || sn->type == STATIC_IPV4_GATEWAY_IFNAME)
+			family = AF_INET;
+		else
+			family = AF_INET6;
+		break;
+
+	case STATIC_IFNAME:
+	case STATIC_BLACKHOLE:
+	default:
+		zlog_err("%s: invalid next hop type", __func__);
+		return;
+	}
+
+	/* Reconfigure or allocate new memory. */
+	if (sn->bsp == NULL)
+		sn->bsp = bfd_sess_new(static_next_hop_bfd_updatecb, sn);
+
+	/* Configure the session. TODO source address.*/
```

Comment:
This TODO must be fixed. Question about the source addresses also in the CLI part.